File: android/app/src/main/java/vn/team/freechat/ContactSearchActivity.java
package vn.team.freechat;

import android.content.Intent;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ListView;

import com.tvd12.ezyfoxserver.client.entity.EzyArray;
import com.tvd12.ezyfoxserver.client.entity.EzyObject;

import java.util.ArrayList;
import java.util.List;

import tvd12.com.ezyfoxserver.client.R;
import vn.team.freechat.adapter.ContactListAdapter;
import vn.team.freechat.adapter.SearchContactListAdapter;
import vn.team.freechat.data.ChannelUsers;
import vn.team.freechat.model.SearchContactModel;
import vn.team.freechat.mvc.IController;
import vn.team.freechat.mvc.IView;
import vn.team.freechat.mvc.Mvc;
import vn.team.freechat.socket.SocketRequests;

/**
 * Created by tavandung12 on 10/1/18.
 */

public class ContactSearchActivity extends AppCompatActivity {

    private String username;
    private IController contactController;
    private IController connectionController;

    private View loadingView;
    private View backButtonView;
    private ListView contactListView;
    private SearchContactListAdapter contactListAdapter;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_contact_search);
        initComponents();
        initViews();
        setViewControllers();
    }

    @Override
    protected void onStart() {
        super.onStart();
        connectionController.addView("show-loading", new IView() {
            @Override
            public void update(Object data) {
                loadingView.setVisibility(View.VISIBLE);
            }
        });
        connectionController.addView("hide-loading", new IView() {
            @Override
            public void update(Object data) {
                loadingView.setVisibility(View.GONE);
            }
        });
        contactController.addView("search-contacts", new IView() {
            @Override
            public void update(Object data) {
                addContacts((EzyArray)data);
            }
        });
    }

    @Override
    protected void onResume() {
        super.onResume();
        sendGetSuggestContacts();
    }

    @Override
    protected void onStop() {
        super.onStop();
        loadingView.setVisibility(View.GONE);
        contactController.removeView("search-contacts");
    }

    private void initComponents() {
        Mvc mvc = Mvc.getInstance();
        username = getIntent().getStringExtra("username");
        contactController = mvc.getController("contact");
        connectionController = mvc.getController("connection");
    }

    private void initViews() {
        loadingView = findViewById(R.id.loading);
        backButtonView = findViewById(R.id.back);
        contactListView = findViewById(R.id.contactList);
        contactListAdapter = new SearchContactListAdapter(this);
        contactListView.setAdapter(contactListAdapter);
    }

    private void setViewControllers() {
        backButtonView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                backToContactView();
            }
        });
        contactListView.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                loadingView.setVisibility(View.VISIBLE);
                ContactListAdapter adapter = (ContactListAdapter) parent.getAdapter();
                onContactItemClick(adapter.getItemModel(position).getContact());
            }
        });
    }

    private void sendGetSuggestContacts() {
        SocketRequests.sendGetSuggestContacts();
    }

    private void addContacts(EzyArray contacts) {
        List<SearchContactModel> models = new ArrayList<>();
        for(int i = 0 ; i < contacts.size() ; ++i) {
            EzyObject item = contacts.get(i, EzyObject.class);
            String username = item.get("username", String.class);
            String fullName = item.get("fullName", String.class, "");
            models.add(new SearchContactModel(username, fullName));
        }
        contactListAdapter.setItemModels(models);
        contactListAdapter.notifyDataSetChanged();
    }

    private void onContactItemClick(ChannelUsers contact) {
        Intent intent = new Intent(this, MessageActivity.class);
        intent.putExtra("targetContact", contact);
        startActivity(intent);
    }

    private void backToContactView() {
        onBackPressed();
    }
}

File: kotlin/app/src/main/java/vn/team/freechat_kotlin/ContactSearchActivity.kt
package vn.team.freechat_kotlin

import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.view.KeyEvent
import android.view.View
import android.view.inputmethod.EditorInfo
import android.widget.EditText
import android.widget.ListView
import android.widget.TextView

import com.tvd12.ezyfoxserver.client.entity.EzyArray

import vn.team.freechat_kotlin.adapter.SearchContactListAdapter
import vn.team.freechat_kotlin.extension.map
import vn.team.freechat_kotlin.model.SearchContactModel
import vn.team.freechat_kotlin.socket.SocketRequests

/**
 * Created by tavandung12 on 10/1/18.
 */

class ContactSearchActivity : AppCompatActivity() {

    private lateinit var contactController: Controller
    private lateinit var connectionController: Controller

    private lateinit var loadingView: View
    private lateinit var backButtonView: View
    private lateinit var searchButtonView: View
    private lateinit var keywordView: EditText
    private lateinit var contactListView: ListView
    private lateinit var contactListAdapter: SearchContactListAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_contact_search)
        initComponents()
        initViews()
        setViewControllers()
    }

    override fun onStart() {
        super.onStart()
        connectionController.addView("show-loading", object: IView {
            override fun update(viewId: String, data: Any?) {
                loadingView.visibility = View.VISIBLE
            }
        })
        connectionController.addView("hide-loading", object: IView {
            override fun update(viewId: String, data: Any?) {
                loadingView.visibility = View.GONE
            }
        })
        contactController.addView("search-contacts", object: IView {
            override fun update(viewId: String, data: Any?) {
                loadingView.visibility = View.GONE
                addContacts(data as EzyArray)
            }
        })

        contactController.addView("add-contacts", "search", object: IView {
            override fun update(viewId: String, data: Any?) {
                backToContactView()
            }
        })
    }

    override fun onResume() {
        super.onResume()
        sendGetSuggestContacts()
    }

    override fun onStop() {
        super.onStop()
        loadingView.visibility = View.GONE
    }

    override fun onDestroy() {
        super.onDestroy()
        contactController.removeView("search-contacts")
        contactController.removeView("add-contacts", "search")
    }

    private fun initComponents() {
        val mvc = Mvc.getInstance()
        contactController = mvc.getController("contact")
        connectionController = mvc.getController("connection")
    }

    private fun initViews() {
        loadingView = findViewById(R.id.loading)
        keywordView = findViewById(R.id.keyword)
        backButtonView = findViewById(R.id.back)
        searchButtonView = findViewById(R.id.search)
        contactListView = findViewById(R.id.contactList)
        contactListAdapter = SearchContactListAdapter(this)
        contactListView.adapter = contactListAdapter
    }

    private fun setViewControllers() {
        backButtonView.setOnClickListener {
            loadingView.visibility = View.GONE
            backToContactView()
        }
        searchButtonView.setOnClickListener {
            loadingView.visibility = View.VISIBLE
            sendSearchContacts()
        }
        keywordView.setOnEditorActionListener(object: TextView.OnEditorActionListener {
            override fun onEditorAction(v: TextView?, actionId: Int, event: KeyEvent?): Boolean {
                if (actionId == EditorInfo.IME_ACTION_SEARCH) {
                    sendSearchContacts()
                    return true
                }
                return false
            }
        })
        contactListView.setOnItemClickListener { parent, _, position, _ ->
            loadingView.visibility = View.VISIBLE
            val adapter = parent.adapter as SearchContactListAdapter
            onContactItemClick(adapter.getItemModel(position).username)
        }
    }

    private fun sendSearchContacts() {
        val keyword = keywordView.text.toString()
        if(keyword.isEmpty())
            SocketRequests.sendGetSuggestContacts()
        else
            SocketRequests.sendSearchContacts(keyword)
    }

    private fun sendGetSuggestContacts() {
        SocketRequests.sendGetSuggestContacts()
    }

    private fun addContacts(contacts: EzyArray) {
        val models = contacts.map {
            SearchContactModel(
                username = it.get<String>("username"),
                fullName = it.get<String>("fullName")
            )
        }
        contactListAdapter.setItemModels(models)
        contactListAdapter.notifyDataSetChanged()
    }

    private fun onContactItemClick(username: String) {
        SocketRequests.sendAddContact(username)
    }

    private fun backToContactView() {
        onBackPressed()
    }
}
